tests/trickle: nondeterministic failure

<!--
If your issue is a usage question, please submit it to the user mailing-list
users@riot-os.org or to the developer mailing-list devel@riot-os.org.
If your issue is related to security, please submit it to the security
mailing-list security@riot-os.org.
-->

#### Description
I've got results very random results with the `trickle` test. This snippet has been reported in at least `samr21-xpro`, `arduino-mega2560`, `iotlab-m3` and `nucleo-l152re`.
This is the kind of output of the test:
```
make RIOT_CI_BUILD=1 CC_NOCOLOR=1 --no-print-directory -C /home/kevinweiss/WorkingDirectory/RIOT/tests/trickle test
ssh -oProxyCommand="ssh weiss@riot-ci.inet.haw-hamburg.de -W %h:%p" -t pi@10.28.48.11 'socat - open:/dev/ttyACM0,b115200,echo=0,raw'
main(): This is RIOT! (Version: buildtest)
[START]
now = 44987, prev_now = 3935, diff = 41052
now = 138904, prev_now = 44987, diff = 93917
now = 380000, prev_now = 138904, diff = 241096
now = 830277, prev_now = 380000, diff = 450277
now = 1648560, prev_now = 830277, diff = 818283
[TRICKLE_RESET]
now = 1666973, prev_now = 1648560, diff = 18413
now = 1694341, prev_now = 1666973, diff = 27368
now = 1772710, prev_now = 1694341, diff = 78369
now = 1845080, prev_now = 1772710, diff = 72370
[FAILURE]
Timeout in expect script at "child.expect(u"now = \d+, prev_now = \d+, diff = \d+")" (tests/trickle/tests/01-run.py:22)

/home/kevinweiss/WorkingDirectory/RIOT/tests/trickle/../../Makefile.include:557: recipe for target 'test' failed
make: *** [test] Error 1

Return value: 2
```

This doesn't seem to be related to #9052 
<!--
Example: Cannot build gnrc_networking application for samr21-xpro board.
-->

#### Steps to reproduce the issue
This bug is hard to reproduce. However, it's run via `make -C tests/trickle flash test`.
<!--
Try to describe as precisely as possible here the steps required to reproduce
the issue. Here you can also describe your hardware configuration, the network
setup, etc.
-->

#### Expected results
<!--
Example: The gnrc_networking application builds on samr21-xpro.
-->
Test passes
#### Actual results
<!--
Please paste or specifically describe the actual output.
-->
Test fails randomly.

--- a/tests/trickle/main.c
+++ b/tests/trickle/main.c
@@ -31,36 +31,37 @@
 #define FIRST_ROUND     (5)
 #define SECOND_ROUND    (12)
 
-static uint32_t prev_now = 0, prev_diff = 0;
+static uint32_t old_t = 0;
 static bool error = false;
+
+static void callback(void *args);
+
+static trickle_t trickle = { .callback = { .func = &callback,
+                                           .args = NULL } };
 
 static void callback(void *args)
 {
     (void) args;
     uint32_t now = xtimer_now_usec();
-    uint32_t diff = (uint32_t) (now - prev_now);
 
-    printf("now = %" PRIu32 ", prev_now = %" PRIu32 ", diff = %" PRIu32
-           "\n", now, prev_now, diff);
+    printf("now = %" PRIu32 ", t = %" PRIu32 "\n", now, trickle.t);
 
-    if (prev_diff >= diff) {
+    /* previous `t` is chosen from a smaller interval [I/2, I).
+     * Current `t` is chosen from interval [I, 2*I).
+     * Hence, `old_t` must be smaller than current `t` */
+    if (old_t >= trickle.t) {
         error = true;
     }
 
-    prev_now = now;
-    prev_diff = diff;
+    old_t = trickle.t;
 
     return;
 }
-
-static trickle_t trickle = { .callback = { .func = &callback,
-                                           .args = NULL } };
 
 int main(void)
 {
     msg_t msg;
     unsigned counter = 0;
-    prev_now = xtimer_now_usec();
 
     trickle_start(sched_active_pid, &trickle, TRICKLE_MSG, TR_IMIN,
                   TR_IDOUBLINGS, TR_REDCONST);
@@ -69,7 +70,7 @@ int main(void)
 
     while (!error) {
         if (counter == FIRST_ROUND) {
-            prev_diff = 0;
+            old_t = 0;
             trickle_reset_timer(&trickle);
             puts("[TRICKLE_RESET]");
         }

--- a/tests/trickle/tests/01-run.py
+++ b/tests/trickle/tests/01-run.py
@@ -14,12 +14,12 @@ def testfunc(child):
     child.expect_exact("[START]")
 
     for i in range(5):
-        child.expect(u"now = \d+, prev_now = \d+, diff = \d+")
+        child.expect(u"now = \\d+, t = \\d+")
 
     child.expect_exact("[TRICKLE_RESET]")
 
     for i in range(7):
-        child.expect(u"now = \d+, prev_now = \d+, diff = \d+")
+        child.expect(u"now = \\d+, t = \\d+")
 
     child.expect_exact("[SUCCESS]")
 

